From 8d3b3fc4774f5112c11f1275f98572a83d0db382 Mon Sep 17 00:00:00 2001
From: Jakub Filak <jfilak@redhat.com>
Date: Thu, 17 Oct 2013 12:47:56 +0200
Subject: [LIBREPORT PATCH] Client API: introduce non-iteractive mode

Avoid blocking waiting for a user response when a program needs some
input from a user but it runs completely without user interaction.

Related to rhbz#1018297

Signed-off-by: Jakub Filak <jfilak@redhat.com>
---
 src/lib/client.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/src/lib/client.c b/src/lib/client.c
index 48e7f3c..62e4fcc 100644
--- a/src/lib/client.c
+++ b/src/lib/client.c
@@ -25,6 +25,11 @@ static int is_slave_mode()
     return getenv("REPORT_CLIENT_SLAVE") != NULL;
 }
 
+static int is_noninteractive_mode()
+{
+    return getenv("REPORT_CLIENT_NONINTERACTIVE") != NULL;
+}
+
 /* Returns 1 if echo has been changed from another state. */
 int set_echo(int enable)
 {
@@ -62,6 +67,13 @@ int ask_yes_no(const char *question)
 
     fflush(stdout);
 
+    if (!is_slave_mode() && is_noninteractive_mode())
+    {
+        putchar('\n');
+        fflush(stdout);
+        return 0;
+    }
+
     char response[16];
     if (NULL == fgets(response, sizeof(response), stdin))
         return 0;
@@ -78,6 +90,13 @@ char *ask(const char *question)
 
     fflush(stdout);
 
+    if (!is_slave_mode() && is_noninteractive_mode())
+    {
+        putchar('\n');
+        fflush(stdout);
+        return NULL;
+    }
+
     char *result = xmalloc_fgets(stdin);
     strtrimch(result, '\n');
 
@@ -93,6 +112,13 @@ char *ask_password(const char *question)
 
     fflush(stdout);
 
+    if (!is_slave_mode() && is_noninteractive_mode())
+    {
+        putchar('\n');
+        fflush(stdout);
+        return NULL;
+    }
+
     bool changed = set_echo(false);
 
     char *result = xmalloc_fgets(stdin);
-- 
1.8.3.1

